content/blog/2022/2022-11-21-building-csv-importer-with-nodejs.md

```diff
@@ -0,0 +1,953 @@
+            accessor: "company"
+          },
+          {
+            Header: "Address",
+            accessor: "address"
+          },
+          {
+            Header: "Location",
+            accessor: "location"
+          },
+          {
+            Header: "Salary",
+            accessor: "salary"
+          },
+          {
+            Header: "Role",
+            accessor: "role"
+          },
+          {
+            Header: "Direct Reportee",
+            accessor: "children",
+            Cell: ({ cell: { value } }) => <Children values={value} />
+          }
+        ]
+      }
+    ],
+    []
+  );
+
+  useEffect(() => {
+    (async () => {
+      const result = await axios("http://localhost:8080/api/employees");
+      setData(result.data);
+    })();
+  }, []);
+
+  const selectFile = (event) => {
+    setSelectedFiles(event.target.files);
+  };
+
+  const upload = () => {
+    let currentFile = selectedFiles[0];
+
+    setProgress(0);
+    setCurrentFile(currentFile);
+
+    uploadToServer(currentFile, (event) => {
+      setProgress(Math.round((100 * event.loaded) / event.total));
+    })
+      .then(async (response) => {
+        setMessage(response.data.message);
+        const result = await axios("http://localhost:8080/api/employees");
+        setData(result.data);
+      })
+      .catch(() => {
+        setProgress(0);
+        setMessage("Could not upload the file!");
+        setCurrentFile(undefined);
+      });
+
+    setSelectedFiles(undefined);
+  };
+
+  return (
+    <div className="App">
+      <div>
+      {currentFile && (
+        <div className="progress">
+          <div
+            className="progress-bar progress-bar-info progress-bar-striped"
+            role="progressbar"
+            aria-valuenow={progress}
+            aria-valuemin="0"
+            aria-valuemax="100"
+            style={{ width: progress + "%" }}
+          >
+            {progress}%
+          </div>
+        </div>
+      )}
+
+      <label className="btn btn-default">
+        <input type="file" onChange={selectFile} />
+      </label>
+
+      <button
+        className="btn btn-success"
+        disabled={!selectedFiles}
+        onClick={upload}
+      >
+        Upload
+      </button>
+
+      <div className="alert alert-light" role="alert">
+        {message}
+      </div>
+    </div>
+      <Table columns={columns} data={data} />
+    </div>
+  );
+}
+
+export default App;
+
+```
+
+We have defined `useMemo` hook to map the attributes from the incoming API data to columns in `react-table` component. We have also defined the `Children` and `Avatar` component to render the images of employees and their children.
+
+Finally, we can run our app by executing the following command:
+
+```bash
+cd client/ && npm run start
+```
+
+This will load the UI and we can upload the CSV file and display our employees in the table:
+
+{{% image alt="Final UI" src="images/posts/nodejs-csv-importer/final_ui.png" %}}
```

Comment:
Maybe move the image of the UI to BEFORE the code examples (and explain it a bit) so that the reader knows what we're building.

Reply:
Moved the image to the initial section and explained about whats coming.